File: R/caracQuantiQuali.R
#######################################################
#                                                     #
#                       Carac Quanti                  #
#                                                     #
#######################################################
require("dplyr")
require("data.table")

var_quanti<-function(myDb,target, varX=NULL,keep.all=F,threshold=0){
  myDb=data.table(myDb)
    # Define numerical values
  if (length(varX)==0) varX=
      colnames(myDb[,sapply(myDb,function(x) {is.numeric(x) & !is.factor(x)}),with=F])
  varX=setdiff(varX,target)
  
  
  myDb <- na.omit(myDb[,c(varX,target),with=F])
  # Just to get rid off that annoying message
  for( col in 1:ncol(myDb)){
    if(class(myDb[[col]]) == 'integer'){
      myDb[, eval(colnames(myDb)[col]) := as.double(get(colnames(myDb)[col]))]
    }
  }
  
  
  # Quantitative analysis
  myDb2<-melt(myDb,id=c(target),measure.vars=varX)
  myDb2[,c("group.average","group.n"):=list(mean(value),.N),by=c(target,"variable")]
  myDb2[,c("overall.average","overall.n","overall.std"):=list(mean(value),.N,sd(value)),by=variable]
  myDb2[,value:=NULL]
  myDb2=unique(myDb2)
  # Probability of having different means
  myDb2[,c("vTest"):=list(
    (group.average-overall.average)/(sqrt(((overall.n-group.n)/
                                            ((overall.n-1)*(group.n))))*overall.std))]
  myDb2[,c("Prob","vStars"):=list(1-pnorm(abs(vTest)),
                                  cut(vTest,breaks=c(-Inf,-5,-2.31,-1.64,-1.28,1.28,1.64,2.31,5,Inf),
                                      labels = c("****|","***|","**|","*|","|","|*","|**","|***","|****"))
  )]
  
  # Filter on prob
  if (threshold!=0){
    myDb2=myDb2[abs(Prob)<threshold]
  }
  #sort
  myDb2=myDb2[order(get(target),vTest)]
  
  #Keep numbers ?
  if (keep.all==F){
    myDb2[,c("overall.n","overall.std"):=NULL]
  }
  
  #prompt
  myDb2[]
}


#######################################################
#                                                     #
#                       Carac quali                   #
#                                                     #
#######################################################


var_quali<-function(myDb, target, varX=NULL,keep.all=F,threshold=0){
  myDb=data.table(myDb)
  # I need all factors to be considered as characters
  for( col in 1:ncol(myDb)){
    if(class(myDb[[col]]) == 'factor'){
      myDb[, eval(colnames(myDb)[col]) := as.character(get(colnames(myDb)[col]))]
    }
  }
  
  # Define char values
  if (length(varX)==0) varX=
    colnames(myDb[,sapply(myDb,function(x) {is.character(x)}),with=F])
  varX=setdiff(varX,target)
  
  
  myDb <- na.omit(myDb[,c(varX,target),with=F])
  # Quantitative analysis
  myDb2<-melt(myDb,id=c(target),measure.vars=varX)
  
  # Basic numbers
  ##  Count number of policies per modality
  myDb2[,nkj:=as.double(.N),by=c(target,"variable","value")]
  ##  Count number of policies per target value
  myDb2[,nk_:=as.double(.N),by=c(target,"variable")]
  ##  Count number of policies per variable modality
  myDb2[,n_j:=as.double(.N),by=c("variable","value")]
  ##  Count total number of policies
  myDb2[,n:=as.double(.N),by=c("variable")]

  # Percentages and vTest
  myDb2[,c("PctIntra","PctTot","PctMod","vTest"):=list(
    nkj/nk_,
    n_j/n,
    nkj/n_j,
    (nkj - n_j/n*nk_)/sqrt((n_j/n*nk_ * (n - nk_)/(n - 1))*(1-n_j/n))
  )]
  
  
  # Indicators
  myDb2=unique(myDb2)
  # Probability of having different means
  myDb2[,c("Prob","vStars"):=list(1-pnorm(abs(vTest)),
                                  cut(vTest,breaks=c(-Inf,-5,-2.31,-1.64,-1.28,1.28,1.64,2.31,5,Inf),
                                      labels = c("****|","***|","**|","*|","|","|*","|**","|***","|****"))
  )]
  
  # Filter on prob
  if (threshold!=0){
    myDb2=myDb2[abs(Prob)<threshold]
  }
  
  # Ranking before sorting
  myRank=rank(myDb2[,max(abs(vTest)),by=variable]$V1)
  myDb2[,variable:=factor(variable,levels=levels(myDb2$variable)[myRank])]
  
  #sort
  myDb2=myDb2[order(get(target),variable,vTest)]
  
  #Keep numbers ?
  if (keep.all==F){
    myDb2[,c("nkj","nk_","n_j","n"):=NULL]
  }
  #prompt
  myDb2[]
}

File: Python/caracQuantiQuali.py
import pandas as pd
import numpy as np
from scipy.stats import norm

def var_quanti(myDb, target, varX=None,keep_all=False,threshold=0):
    if (varX == None):
        varX = myDb.select_dtypes(include=[np.number]).columns.tolist()
        # varX = [c for c in varX if len(myDb[c].unique())<20]
    varX = list(set(varX) - set([target]))

    melted = pd.melt(myDb, id_vars=target, value_vars=varX)
    melted = melted.assign(N_=1)

    melted['group_average'] = melted.groupby([target, "variable"])['value'].transform(np.mean)
    melted['group_N'] = melted.groupby([target, "variable"])['N_'].transform(sum)
    melted['overall_average'] = melted.groupby(["variable"])['value'].transform(np.mean)
    melted['overall_N'] = melted.groupby(["variable"])['N_'].transform(sum)
    melted['overall_std'] = melted.groupby(["variable"])['value'].transform(np.std)
    melted.drop('value',axis=1,inplace=True)
    melted.drop_duplicates(inplace=True)
    myDb = melted.reset_index(drop=True)

    # Computing vTest
    myDb['vTest'] = ((myDb['group_average'] - myDb['overall_average']) / (
            np.sqrt(
                ((myDb['overall_N'] - myDb['group_N']) / ((myDb['overall_N'] - 1) * (myDb['group_N'])))) *
            myDb['overall_std']))

    # Computing Prob
    myDb['Prob'] = 1 - norm.cdf(abs(myDb['vTest']))

    # displaying stars
    myDb['vStars'] = pd.cut(myDb['vTest'],
                            bins=[-np.float("Inf"), -5, -2.31, -1.64, -1.28, 1.28, 1.64, 2.31, 5, np.float('Inf')],
                            labels=["****|", "***|", "**|", "*|", "|", "|*", "|**", "|***", "|****"])

    # Filtering on probability
    if (threshold > 0):
        myDb[abs(myDb['Prob']) < threshold]

    myDb.sort_values([target, 'vTest'])

    # Dropping useless
    if (keep_all):
        myDb.drop(['overall_N', 'overall_std'], axis=1, inplace=True)

    return myDb


def var_quali(myDb, target, varX=None,keep_all=False,threshold=0):
    if (varX == None):
        varX = db.select_dtypes(exclude=[np.number]).columns.tolist()
        # varX = [c for c in varX if len(myDb[c].unique())<20]
    varX = list(set(varX) - set([target]))

    melted = pd.melt(myDb, id_vars=target, value_vars=varX)
    melted = melted.assign(N_=1)

    # Basic numbers
    ## Count number of policies per modality
    melted['nkj'] = melted.groupby([target, "variable", "value"])['N_'].transform(sum)
    ## Count number of policies per target value
    melted['nk_'] = melted.groupby([target, "variable"])['N_'].transform(sum)
    ## Count number of policies per variable modality
    melted['n_j'] = melted.groupby(["variable", "value"])['N_'].transform(sum)
    ## Count total number of policies
    melted['n'] = melted.groupby(["variable"])['N_'].transform(sum)

    # Frequencies for Chi-2
    melted["PctIntra"] = melted['nkj'] / melted['nk_']
    melted["PctTot"] = melted['n_j'] / melted['n']
    melted["PctMod"] = melted['nkj'] / melted['n_j']
    melted["vTest"] = (melted['nkj'] - melted['n_j'] / melted['n'] * melted['nk_']) / np.sqrt(
        (melted['n_j'] / melted['n'] * melted['nk_'] *
         (melted['n'] - melted['nk_']) / (melted['n'] - 1)) * (1 - melted['n_j'] / melted['n'])
    )
    myDb = melted.drop_duplicates().copy()
    # Indicators :
    # Computing Prob
    myDb['Prob'] = 1 - norm.cdf(abs(myDb['vTest']))

    # displaying stars
    myDb['vStars'] = pd.cut(myDb['vTest'],
                            bins=[-np.float("Inf"), -5, -2.31, -1.64, -1.28, 1.28, 1.64, 2.31, 5, np.float('Inf')],
                            labels=["****|", "***|", "**|", "*|", "|", "|*", "|**", "|***", "|****"])

    # Filter on prob
    if (threshold != 0):
        myDb = myDb[abs(Prob) < threshold]

    # Sorting
    myDb = myDb.sort_values([target, 'variable', 'vTest'])
    if (keep_all == False):
        myDb = myDb.drop(["nkj", "nk_", "n_j", "n"], axis=1)
    return myDb

